src/main/kotlin/org/rust/lang/core/crate/impl/CrateGraphServiceImpl.kt

```diff
@@ -0,0 +1,400 @@
+            pkg.pkg.libTarget?.let {
+                libCrate.cargoTarget = it
+                libCrate.cargoProject = pkg.project
+            }
+        }
+        for (crate in state.nonLibraryCrates) {
+            val newTarget = pkg.pkg.targets.find { it.name == crate.cargoTarget.name }
+                ?: continue
+            crate.cargoTarget = newTarget
+            crate.cargoProject = pkg.project
+        }
+    }
+
+    private data class LoweredPackageDependencies(
+        val buildDeps: List<Crate.Dependency>,
+        val normalAndNonCyclicDevDeps: List<Crate.Dependency>,
+        val cyclicDevDependencies: List<CargoWorkspace.Dependency>
+    )
+
+    private fun lowerPackageDependencies(pkg: ProjectPackage): LoweredPackageDependencies {
+        return when (val dependencies = pkg.pkg.dependencies.classify()) {
+            is SpitDependencies.Classified -> {
+                val buildDeps = lowerDependencies(dependencies.build, pkg)
+                val normalDeps = lowerDependencies(dependencies.normal, pkg)
+                val (nonCyclicDevDeps, cyclicDevDependencies) = lowerDependenciesWithCycles(dependencies.dev, pkg)
+                val normalAndNonCyclicDevDeps = normalDeps + nonCyclicDevDeps
+                LoweredPackageDependencies(buildDeps, normalAndNonCyclicDevDeps, cyclicDevDependencies)
+            }
+            is SpitDependencies.Unclassified -> {
+                // Here we can't distinguish normal/dev/build dependencies because of old Cargo (prior to `1.41.0`).
+                //
+                val (lowered, cyclic) = lowerDependenciesWithCycles(dependencies.dependencies, pkg)
+                LoweredPackageDependencies(lowered, lowered, cyclic)
+            }
+        }
+    }
+
+    private sealed class SpitDependencies {
+        data class Classified(
+            val normal: Collection<CargoWorkspace.Dependency>,
+            val dev: Collection<CargoWorkspace.Dependency>,
+            val build: Collection<CargoWorkspace.Dependency>
+        ) : SpitDependencies()
+
+        // For old Cargo versions prior to `1.41.0`
+        data class Unclassified(val dependencies: Collection<CargoWorkspace.Dependency>) : SpitDependencies()
+    }
+
+    private fun Collection<CargoWorkspace.Dependency>.classify(): SpitDependencies {
+        val unclassified = mutableListOf<CargoWorkspace.Dependency>()
+        val normal = mutableListOf<CargoWorkspace.Dependency>()
+        val dev = mutableSetOf<CargoWorkspace.Dependency>()
+        val build = mutableListOf<CargoWorkspace.Dependency>()
+
+        for (dependency in this) {
+            val visitedDepKinds = enumSetOf<CargoWorkspace.DepKind>()
+
+            for (depKind in dependency.depKinds) {
+                // `cargo metadata` sometimes duplicate `depKinds` for some reason
+                if (!visitedDepKinds.add(depKind.kind)) continue
+
+                when (depKind.kind) {
+                    CargoWorkspace.DepKind.All -> {
```

Comment:
As I understand, `All` is only possible to stdlib dependencies? if yes, maybe it's worth to rename to `Stdlib`?